authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,147 @@
 #  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  * See the License for the specific language governing permissions and
 #  * limitations under the License.
-
+import logging
 
 import ldap
 from flask_securest.authentication_providers.abstract_authentication_provider \
     import AbstractAuthenticationProvider
 from flask_securest import utils
+from flask_securest import constants
+from flask_securest.exceptions import AuthenticationException
+
+DN_ATTRIBUTE_NAME = 'DN'
+DEFAULT_AUTH_ATTR_NAME = 'uid'
 
 
 class LDAPAuthenticationProvider(AbstractAuthenticationProvider):
-    def __init__(self, directory_url):
-        self.directory_url = directory_url
+    def __init__(self,
+                 ldap_url,
+                 base_dn=None,
+                 admin_user_id=None,
+                 admin_password=None,
+                 user_id_attribute=None):
+        self.lgr = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+        self.ldap_url = ldap_url
+        self.base_dn = base_dn
+        self.admin_user_id = admin_user_id
+        self.admin_password = admin_password
+        self.user_id_attribute = user_id_attribute
 
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_subtree_authentication_method():
```

Comment:
1. There is actually a bigger difference here. It's true that both use bind to authenticate but in the subtree search, the configuration has to hold an admin user and admin password whereas the second method does not require these credentials. What you are suggesting is not possible since searching the tree is only possible if the admin username and password were indeed provided in the configuration.. hence self.is_subtree_authentication_method()..
2. If whoever configured the ldap in the blueprint specified a base_dn, it means no one outside this subtree is allowed to login. If a base_dn was not specified then we indeed allow users from the ldap root to login. If the guy configuring the ldap security in the blueprint wants to allow identification using attributes, he should specify the root as a base DN before bootstrapping